python/GafferUI/SpreadsheetUI/_ClipboardAlgo.py

```diff
@@ -206,330 +166,235 @@ def createPlugMatrixFromCells( cellPlugs ) :
-			valueData = data[ "value" ]
-			if "enabled" in nameValuePlug :
-				if "enabled" in data and not ValueAdaptor.canSet( data[ "enabled" ], nameValuePlug[ "enabled" ] ) :
-					return False
+	def settable( p ) :
+		if Gaffer.Animation.isAnimated( p ) :
+			curve = Gaffer.Animation.acquire( p )
+			return not Gaffer.MetadataAlgo.readOnly( curve )
 		else :
-			# Allow simple data to be pasted onto the value plug
-			valueData = data
-
-		return ValueAdaptor.canSet( valueData, nameValuePlug[ "value" ] )
-
-	@classmethod
-	def _set( cls, data, nameValuePlug, atTime ) :
-
-		# We should _never_ set the name of an NVP as it's not exposed in the
-		# UI. It's only there as it simplifies things greatly if we don't
-		# special case plugs in the general case in the Spreadsheet. If we did,
-		# and you pasted across columns, it would duplicate the plug name,
-		# which can have catastrophic results for nodes such as StandardOptions.
-
-		if isinstance( data, IECore.CompoundData ) and "value" in data :
-			valueData = data[ "value" ]
-			enabledData = data[ "enabled" ] if "enabled" in data else None
-		else :
-			# Allow simple data to be pasted onto the value plug
-			valueData = data
-			enabledData = None
-
-		ValueAdaptor.set( valueData, nameValuePlug[ "value" ], atTime )
-
-		if "enabled" in nameValuePlug and enabledData is not None :
-			ValueAdaptor.set( enabledData, nameValuePlug[ "enabled" ], atTime )
-
-ValueAdaptor.registerAdaptor( Gaffer.NameValuePlug, NameValuePlugValueAdaptor )
+			return p.settable()
 
-class CellPlugValueAdaptor( ValueAdaptor ) :
-
-	@classmethod
-	def _canSet( cls, data, cellPlug ) :
-
-		enabledData, valueData = CellPlugValueAdaptor.__enabledAndValueData( data )
+	if Gaffer.MetadataAlgo.readOnly( plug ) or not settable( plug ) :
+		return False
 
-		if enabledData is not None and not ValueAdaptor.canSet( enabledData, cellPlug.enabledPlug() ) :
+	for p in Gaffer.Plug.RecursiveRange( plug ) :
+		if Gaffer.MetadataAlgo.getReadOnly( p ) :
 			return False
 
-		return ValueAdaptor.canSet( valueData, cellPlug[ "value" ] )
-
-	@classmethod
-	def _set( cls, data, cellPlug, atTime ) :
-
-		enabledData, valueData = CellPlugValueAdaptor.__enabledAndValueData( data )
-
-		ValueAdaptor.set( valueData, cellPlug[ "value" ], atTime )
-
-		# Set enabled state last, such that when copying from a cell that doesn't adopt
-		# an enabled plug, to one that does, the final 'enabled' state matches.
-		if enabledData is not None :
-			ValueAdaptor.set( enabledData, cellPlug.enabledPlug(), atTime )
-
-	@staticmethod
-	def __enabledAndValueData( data ) :
-
-		enabledData = None
-
-		if isinstance( data, IECore.CompoundData ) and "value" in data :
-			valueData = data[ "value" ]
-			if "enabled" in data :
-				enabledData = data[ "enabled" ]
-			elif "enabled" in valueData :
-				enabledData = valueData[ "enabled" ]
-		else :
-			valueData = data
+	return True
 
-		return enabledData, valueData
+def __isPlugMatrix( plugMatrix ) :
```

Comment:
I might be missing something, but I don't know why `__isPlugMatrix()` is necessary. When do we pass `plugMatrix` arguments that might not actually be plug matrices? And if we do, might it be better not to pass them in the first place?

Reply:
This mostly just satisfies some of the pre-existing tests for malformed plugMatrices, and enforces that the plugMatrix contains a consistent number of columns per row. I've squashed a todo into e872891 suggesting that this could be removed in the future.